### PR TITLE
Add backend test script

### PR DESCRIPTION
**Pull request by tharikaGitHub**
## Purpose
This PR adds the backend performance testing scripts.

### CODE REVIEW

File: distribution/scripts/jmeter/backend-test.jmx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        <longProp name="ThreadGroup.end_time">1499334866000</longProp>
+        <boolProp name="ThreadGroup.scheduler">true</boolProp>
+        <stringProp name="ThreadGroup.duration">${__P(duration)}</stringProp>
+        <stringProp name="ThreadGroup.delay"></stringProp>
+      </ThreadGroup>
+        <hashTree/>
+        <HeaderManager guiclass="HeaderPanel" testclass="HeaderManager" testname="HTTP Headers" enabled="true">
+          <collectionProp name="HeaderManager.headers">
+            <elementProp name="" elementType="Header">
+              <stringProp name="Header.name">Authorization</stringProp>
```
Comment (chrishantha) [inline distribution/scripts/jmeter/backend-test.jmx]: Remove this header.

File: distribution/scripts/jmeter/run-backend-tests.sh [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+script_dir=$(dirname "$0")
+# Execute common script
+. $script_dir/perf-test-common.sh
+
+function initialize() {
+    export backend_host=$(get_ssh_hostname $backend_ssh_host)
+}
+export -f initialize
+
+declare -A test_scenario0=(
```
Comment (chrishantha) [inline distribution/scripts/jmeter/run-backend-tests.sh:31]: Add HTTPS scenario as well.

File: distribution/scripts/jmeter/run-backend-tests.sh [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    [name]="backend"
+    [display_name]="Backend"
+    [description]="Directly invoking the backend service"
```
Comment (chrishantha) [inline distribution/scripts/jmeter/run-backend-tests.sh]: Use a better name, display name and description

File: distribution/scripts/jmeter/run-backend-tests.sh [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    [skip]=false
+)
+
+function before_execute_test_scenario() {
+    local service_path=${scenario[path]}
+    local protocol=${scenario[protocol]}
+    jmeter_params+=("host=$backend_host" "port=8688" "path=$service_path")
+    jmeter_params+=("payload=$HOME/${msize}B.json" "response_size=${msize}B" "protocol=$protocol")
+    echo "Starting backend service"
+    ssh $backend_ssh_host "./netty-service/netty-start.sh -m $heap"
```
Comment (chrishantha) [inline distribution/scripts/jmeter/run-backend-tests.sh]: Do we need this?

**Review by chrishantha — CHANGES_REQUESTED** [review]

File: distribution/scripts/jmeter/run-backend-tests.sh [hunk 5]
```diff
@@ -29,23 +29,32 @@ function initialize() {
...
-    [name]="backend"
-    [display_name]="Backend"
-    [description]="Directly invoking the backend service"
+    [name]="netty-backend-http"
+    [display_name]="Netty-Backend-HTTP"
+    [description]="This test directly invokes the Netty Backend Service through HTTP protocol"
     [jmx]="backend-test.jmx"
     [protocol]="http"
     [path]="/echo/1.0.0"
     [use_backend]=false
```
Comment (chrishantha) [inline distribution/scripts/jmeter/run-backend-tests.sh]: `use_backend` should be `true` for all scenarios

File: distribution/scripts/jmeter/run-backend-tests.sh [hunk 6]
```diff
@@ -29,23 +29,32 @@ function initialize() {
...
 )
 
+declare -A test_scenario1=(
+    [name]="netty-backend-https"
+    [display_name]="Netty-Backend-HTTPS"
+    [description]="This test directly invokes the Netty Backend Service through HTTPS protocol"
+    [jmx]="backend-test.jmx"
+    [protocol]="https"
+    [path]="/echo/1.0.0"
+    [use_backend]=false
```
Comment (chrishantha) [inline distribution/scripts/jmeter/run-backend-tests.sh]: https scenario needs to have following option:  `[backend_flags]="--enable-ssl"`

File: distribution/scripts/jmeter/run-backend-tests.sh [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+)
+
+function before_execute_test_scenario() {
+    local service_path=${scenario[path]}
+    local protocol=${scenario[protocol]}
+    jmeter_params+=("host=$backend_host" "port=8688" "path=$service_path")
+    jmeter_params+=("payload=$HOME/${msize}B.json" "response_size=${msize}B" "protocol=$protocol")
+}
+
+function after_execute_test_scenario() {
```
Comment (chrishantha) [inline distribution/scripts/jmeter/run-backend-tests.sh:68]: I think that this function is not required.

**Review by chrishantha — CHANGES_REQUESTED** [review]

File: distribution/scripts/jmeter/run-backend-tests.sh (hunk 7 above)
Comment (tharikaGitHub) [reply distribution/scripts/jmeter/run-backend-tests.sh:68]: This function is being referred from perf-test-common script. Cannot remove it. Hence added this method as an empty method returning true.

Comment (chrishantha) [reply distribution/scripts/jmeter/run-backend-tests.sh:68]: That's okay. I think you can just add `return 0`, which is more clear than using `:`

File: distribution/scripts/jmeter/run-backend-tests.sh [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+}
+export -f initialize
+
+declare -A test_scenario0=(
+    [name]="netty-backend-http"
+    [display_name]="Netty Backend HTTP"
+    [description]="This test directly invokes the Netty Backend Service through HTTP protocol"
+    [jmx]="backend-test.jmx"
+    [protocol]="http"
+    [path]="/echo/1.0.0"
```
Comment (chrishantha) [inline distribution/scripts/jmeter/run-backend-tests.sh]: Let's just add "/" for path. There is no context in Netty Service.

File: distribution/scripts/jmeter/run-backend-tests.sh [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+function initialize() {
+    export backend_host=$(get_ssh_hostname $backend_ssh_host)
+}
+export -f initialize
+
+declare -A test_scenario0=(
+    [name]="netty-backend-http"
+    [display_name]="Netty Backend HTTP"
+    [description]="This test directly invokes the Netty Backend Service through HTTP protocol"
```
Comment (chrishantha) [inline distribution/scripts/jmeter/run-backend-tests.sh]: [description]="An HTTP echo service implemented in Netty"

File: distribution/scripts/jmeter/run-backend-tests.sh [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    [protocol]="http"
+    [path]="/echo/1.0.0"
+    [use_backend]=true
+    [skip]=false
+)
+
+declare -A test_scenario1=(
+    [name]="netty-backend-https"
+    [display_name]="Netty Backend HTTPS"
+    [description]="This test directly invokes the Netty Backend Service through HTTPS protocol"
```
Comment (chrishantha) [inline distribution/scripts/jmeter/run-backend-tests.sh]: [description]="An HTTPS echo service implemented in Netty"

File: distribution/scripts/jmeter/run-backend-tests.sh [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    [jmx]="backend-test.jmx"
+    [protocol]="http"
+    [path]="/echo/1.0.0"
+    [use_backend]=true
+    [skip]=false
+)
+
+declare -A test_scenario1=(
+    [name]="netty-backend-https"
+    [display_name]="Netty Backend HTTPS"
```
Comment (chrishantha) [inline distribution/scripts/jmeter/run-backend-tests.sh]: [display_name]="HTTPS echo service"

File: distribution/scripts/jmeter/run-backend-tests.sh [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+. $script_dir/perf-test-common.sh
+
+function initialize() {
+    export backend_host=$(get_ssh_hostname $backend_ssh_host)
+}
+export -f initialize
+
+declare -A test_scenario0=(
+    [name]="netty-backend-http"
+    [display_name]="Netty Backend HTTP"
```
Comment (chrishantha) [inline distribution/scripts/jmeter/run-backend-tests.sh]: [display_name]="HTTP echo service"

File: distribution/scripts/jmeter/run-backend-tests.sh [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+)
+
+declare -A test_scenario1=(
+    [name]="netty-backend-https"
+    [display_name]="Netty Backend HTTPS"
+    [description]="This test directly invokes the Netty Backend Service through HTTPS protocol"
+    [jmx]="backend-test.jmx"
+    [protocol]="https"
+    [backend_flags]="--enable-ssl"
+    [path]="/echo/1.0.0"
```
Comment (chrishantha) [inline distribution/scripts/jmeter/run-backend-tests.sh]: [path]="/"

**Review by chrishantha — CHANGES_REQUESTED** [review]

File: distribution/scripts/cloudformation/run-backend-tests.sh [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+# under the License.
+#
+# ----------------------------------------------------------------------------
+# Run backend performance tests on AWS Cloudformation Stacks
+# ----------------------------------------------------------------------------
+
+export script_name="$0"
+export script_dir=$(dirname "$0")
+
+export backend_ec2_instance_type=""
```
Comment (chrishantha) [inline distribution/scripts/cloudformation/run-backend-tests.sh]: Just noticed. This is not required.

File: distribution/scripts/cloudformation/run-backend-tests.sh [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+export aws_cloudformation_template_filename="backend_perf_test_cfn.yaml"
+export application_name="Backend Server"
+export metrics_file_prefix="netty"
+
+function usageCommand() {
```
Comment (chrishantha) [inline distribution/scripts/cloudformation/run-backend-tests.sh]: Not required

File: distribution/scripts/cloudformation/run-backend-tests.sh [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+    echo "-A <backend_ec2_instance_type>"
+}
+export -f usageCommand
+
+function usageHelp() {
```
Comment (chrishantha) [inline distribution/scripts/cloudformation/run-backend-tests.sh]: Not required

File: distribution/scripts/cloudformation/run-backend-tests.sh [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+    echo "-A: Amazon EC2 Instance Type for Backend."
+}
+export -f usageHelp
+
+while getopts ":f:d:k:n:j:o:g:s:b:r:J:S:N:t:p:w:hA:" opt; do
```
Comment (chrishantha) [inline distribution/scripts/cloudformation/run-backend-tests.sh]: Not required

File: distribution/scripts/cloudformation/run-backend-tests.sh [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+        ;;
+    *)
+        opts+=("-${opt}")
+        [[ -n "$OPTARG" ]] && opts+=("$OPTARG")
+        ;;
+    esac
+done
+shift "$((OPTIND - 1))"
+
+function validate() {
```
Comment (chrishantha) [inline distribution/scripts/cloudformation/run-backend-tests.sh]: Not required

File: distribution/scripts/cloudformation/run-backend-tests.sh [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+    if [[ -z $backend_ec2_instance_type ]]; then
+        echo "Please provide the Amazon EC2 Instance Type for Backend."
+        exit 1
+    fi
+}
+export -f validate
+
+function get_test_metadata() {
+    echo "application_name=$application_name"
+    echo "backend_ec2_instance_type=$backend_ec2_instance_type"
```
Comment (chrishantha) [inline distribution/scripts/cloudformation/run-backend-tests.sh]: Not required

File: distribution/scripts/cloudformation/run-backend-tests.sh [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+}
+export -f get_test_metadata
+
+function get_cf_parameters() {
+    echo "BackendInstanceType=$backend_ec2_instance_type"
```
Comment (chrishantha) [inline distribution/scripts/cloudformation/run-backend-tests.sh]: Not required

**Review by chrishantha — CHANGES_REQUESTED** [review]

File: distribution/scripts/jmeter/template/summary.md [hunk 21]
```diff
@@ -0,0 +1,59 @@
+# {{ parameters.application_name }} Performance Test Results
```
Comment (chrishantha) [inline distribution/scripts/jmeter/template/summary.md]: Directory should be `templates`

**Review by chrishantha — CHANGES_REQUESTED** [review]

File: distribution/scripts/cloudformation/run-backend-tests.sh [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+    echo "Heap Size"
+    echo "Concurrent Users"
+    echo "Message Size (Bytes)"
+    echo "Back-end Service Delay (ms)"
+    echo "Error %"
+    echo "Throughput (Requests/sec)"
+    echo "Average Response Time (ms)"
+    echo "Standard Deviation of Response Time (ms)"
+    echo "99th Percentile of Response Time (ms)"
+    echo "Backend GC Throughput (%)"
```
Comment (chrishantha) [inline distribution/scripts/cloudformation/run-backend-tests.sh]: Here the name should be the `application_name`

**Review by chrishantha — CHANGES_REQUESTED** [review]

**Review by chrishantha — APPROVED** [review]

**Pull request merged**